gcloud/datastore/client.py

```diff
@@ -180,14 +181,15 @@ class Client(_BaseClient):
 
     def __init__(self, project=None, namespace=None,
                  credentials=None, http=None):
-        project = _determine_default_project(project)
-        if project is None:
-            raise EnvironmentError('Project could not be inferred.')
-        self.project = project
+        _ClientProjectMixin.__init__(self, project=project)
         self.namespace = namespace
         self._batch_stack = _LocalStack()
         super(Client, self).__init__(credentials, http)
 
+    @staticmethod
+    def _determine_default(project):
+        return _determine_default_project(project)
```

Comment:
I guess I had in mind a frameworky bit where the `_ClientProjectMixin` would use a class-attribute of the environment variable names to check, defaulting to `[gcloud.environment_vars.PROJECT]`, but allowing the datastore and pubsub clients to add the appropriate varnames for their emulators:  then the whole dance would be in one place, instead of scattered across three packages / four modules.